Resources/Storage/StorageElement.py

```diff
@@ -732,6 +753,95 @@ def __getattr__( self, name ):
       return self.__executeMethod
 
     raise AttributeError( "StorageElement does not have a method '%s'" % name )
+  
+
+      
+
+  def addAccountingOperation( self, lfns, startDate, elapsedTime, storageParameters, callRes ):
+    """
+        Generates a DataOperation accounting if needs to be, and adds it to the DataStore client cache
+
+        :param lfns : list of lfns on which we attempted the operation
+        :param startDate : datetime, start of the operation
+        :param elapsedTime : time (seconds) the operation took
+        :param storageParameters : the parameters of the plugins used to perform the operation
+        :param callRes : the return of the method call, S_OK or S_ERROR
+
+        The operation is generated with the OperationType "se.methodName"
+        The TransferSize and TransferTotal for directory methods actually take into
+        account the files inside the directory, and not the amount of directory given
+        as parameter
+
+
+    """
+  
+    if self.methodName not in ( self.readMethods + self.writeMethods + self.removeMethods ):
+      return
+  
+    baseAccountingDict = {}
+    baseAccountingDict['OperationType'] = 'se.%s' % self.methodName
+    baseAccountingDict['User'] = getProxyInfo().get( 'Value', {} ).get( 'username', 'unknown' )
+    baseAccountingDict['RegistrationTime'] = 0.0
+    baseAccountingDict['RegistrationOK'] = 0
+    baseAccountingDict['RegistrationTotal'] = 0
+    baseAccountingDict['Destination'] = self.name
+    baseAccountingDict['Source'] = siteName()
+    baseAccountingDict['TransferTotal'] = 0
+    baseAccountingDict['TransferOK'] = 0
+    baseAccountingDict['TransferSize'] = 0
+    baseAccountingDict['TransferTime'] = 0.0
+    baseAccountingDict['FinalStatus'] = 'Successful'
+
+    oDataOperation = DataOperation()
+    oDataOperation.setValuesFromDict( baseAccountingDict )
+    oDataOperation.setStartTime( startDate )
+    oDataOperation.setEndTime( startDate + datetime.timedelta( seconds = elapsedTime ) )
+    oDataOperation.setValueByKey( 'TransferTime', elapsedTime )
+    oDataOperation.setValueByKey( 'Protocol', storageParameters.get( 'Protocol', 'unknown' ) )
+  
+    if not callRes['OK']:
+      # Everything failed
+      oDataOperation.setValueByKey( 'TransferTotal', len( lfns ) )
+      oDataOperation.setValueByKey( 'FinalStatus', 'Failed' )
+    else:
+      totalSize = 0
+
+      oDataOperation.setValueByKey( 'TransferTotal', len( callRes['Value']['Successful'] ) )
```

Comment:
Must be damned sure that all SE implementations will return the correct structure!